proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BackendPlaySessionHandler.java

```diff
@@ -173,12 +179,72 @@ public boolean handle(AvailableCommands commands) {
       }
     }
 
-    server.getEventManager().fire(
-        new PlayerAvailableCommandsEvent(serverConn.getPlayer(), rootNode))
-        .thenAcceptAsync(event -> playerConnection.write(commands), playerConnection.eventLoop());
+    server
+        .getEventManager()
+        .fire(new PlayerAvailableCommandsEvent(serverConn.getPlayer(), rootNode))
+        .thenAcceptAsync(
+            event -> {
+              commands.setRootNode(
+                  (RootCommandNode<CommandSource>) filterNode(rootNode, new IdentityHashMap<>()));
+              playerConnection.write(commands);
+            },
+            playerConnection.eventLoop());
     return true;
   }
 
+  /**
+   * Creates a deep copy of the provided command node, but removes any node that are not accessible
+   * by the player (respecting the requirement of the node).
+   *
+   * @param source source node
+   * @param nodeMapping mapped nodes
+   * @return filtered node
+   */
+  private CommandNode<CommandSource> filterNode(
```

Comment:
This only works for `BrigadierCommand`s, I couldn't think of a way to implement it for simple and raw commands that expect a Context object (creating the context manually is hacky at best)

Reply:
Sorry if I'm missing something, but why do we need to treat redirects as a special case? See `BrigadierUtils` for a deep clone method example.

Reply:
I can also implement it on the simple/raw commands, it is just command completion, but this method is only for `BrigadierCommand`s

Reply:
> why do we need to treat redirects as a special case

We need to filter the redirects too

Reply:
The problem is simple and raw commands don't use the requires predicate since it doesn't pass the used command and thus the context object cannot be created.

Reply:
> since it doesn't pass the used command

wdym

Reply:
See https://github.com/VelocityPowered/Velocity/blob/76173e414506f361c5acc60318a16bcede44edbb/proxy/src/main/java/com/velocitypowered/proxy/command/CommandNodeFactory.java#L36

Reply:
Can't we make the command to forward only on failure? So that we can utilise .requires for better experience?
(asking u before touching anything of that code, I'm scared 👀)